Refactor Mixture distribution for V4

ricardoV94 opened the pull request:
This PR is an attempt to refactor (Marginalized) Mixture distributions for V4

- [x] Refactor Mixture with iterable components
- [x] Refactor Mixture with single component
- [x] Refactor NormalMixture
- [x] Rebase from main and handle new meaning of size for multivariate dists
- [x] Reenable TestMixtureVsLatent
- [x] Deprecate MixtureSameFamily (keep tests)
- [x] Add more design decision context to refactor commit
- [x] Add moment and tests
- [x] Allow for Nested Mixtures / Other symbolic distributions as components (reminder #5533)

## Changes
There are two big changes in how Mixture works compared to V3:
1. The support dimensionality of the Mixture no longer depends on the size of the weights, but on the support dimensionality of the components. This means one cannot use a batched scalar distribution (like Normal) as if it was a vector distribution, which was supposedly allowed by the docstrings. However, while the old random method respected this, the logp method did not (the logp would be indifferent to mixing of values across this "fake" vector components), so this is not clearly a regression.
```python
# This mixture now assumes interchangeability across the 5 values of each component, 
# regardless of the dimensionality of the weights
mix = pm.Mixture.dist(w=[0.5, 0.5], comp_dists=pm.Normal.dist([-10, 10], size=(5, 2)))
```
I am exploring adding a keyword argument to override the support dimensionality in a follow-up PR. I managed to do this for the random method, but haven't had time yet to figure out what needs to be changed in the logp method.

2. Nested Mixtures or Mixtures using non pure RandomVariables (such as Censored variables) are not yet possible. This will require creating a dispatch to retrieve and manipulate these variables (i.e., check ndim_supp, resize), in the same way that's being done with the component RandomVariables. This should be straightforward, since these had to be implemented as methods of the SymbolicDistribution anyway...

Closes #4781

larryshamalama committed:
Add Mixture moments

ricardoV94 committed:
Refactor Mixture distribution

Mixtures now use an `OpFromGraph` that encapsulates the Aesara random method. This is used so that logp can be easily dispatched to the distribution without requiring involved pattern matching. The Mixture random and logp methods now fully respect the support dimensionality of its components, whereas previously only the logp method did, leading to inconsistencies between the two methods.

In the case where the weights (or size) indicate the need for more draws than what is given by the component distributions, the latter are resized to ensure there are no repeated draws.

This refactoring forces Mixture components to be basic RandomVariables, meaning that nested Mixtures or Mixtures of Symbolic distributions (like Censored) are not currently possible.

Co-authored-by: Larry Dong <larry.dong@mail.utoronto.ca>

ricardoV94 committed:
Move NormalMixture tests to their own class

ricardoV94 committed:
Use specific imports in test_mixture

ricardoV94 committed:
Move mixture random tests from test_distributions_random to test_mixture

ricardoV94 commented:
@Sayam753 could you give some context on what was the idea with `MixtureSameFamily`?

Now that we have meta information about the components (ndim_supp, ndims_params), it is perhaps no longer needed?

ricardoV94 committed:
Reenable Mixture tests in pytest workflow

Sayam753 commented:
> could you give some context on what was the idea with MixtureSameFamily?

`MixtureSameFamily` distribution helps to create mixture distribution for multivariate distributions.

The legacy `Mixture` distribution assumes that the mixture components are present in the last dimension of a distribution. This creates a problem for multivariate distributions, because in this case, the last dimension corresponds to events, not the mixture components.

`MixtureSameFamily` takes an input an integer asking which axis to consider for mixture components. And it reduces that axis during logp computations.

@lucianopaz is the magician behind this distribution.

ricardoV94 commented:
If I understand correctly, I think we can infer the correct mixture axis from the meta information we have now. 

That's my TODO point above. So far, I have just been hacking it by trial and error. 

Do you think there is inherent ambiguity still?  Even considering ndim_supp, ndims_params, shape_from_params... and whatever else methods we have to reason about shapes of RandomVariables?

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
-            self.comp_is_distribution = True
+        stacked_components_ = at.stack(components_type, axis=-1)
+        if weights.ndim < stacked_components_.ndim:
+            weights_ = at.shape_padaxis(weights_type, axis=weights_type.ndim - 1)
```
This is very likely completely wrong. I just trial and errored to get the tests to pass. From what @Sayam753 said in https://github.com/pymc-devs/pymc/pull/5438#issuecomment-1030590138, it should probably be based on the support dimensionality of the components, not on the dimensionality of the weights

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
-        """
-        w = self.w
+    # TODO: Is one padding always enough?
+    if weights.ndim < components_logp.ndim:
```
Gotta figure out how weights should be padded in the logp

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
+@_get_moment.register(MarginalMixtureRV)
+def get_moment_marginal_mixture(op, rv, rng, weights, *components):
+    moment_components = at.stack([get_moment(component) for component in components], axis=-1)
+    return at.sum(weights * moment_components, axis=-1)
```
Needs padding of weights as well

larryshamalama commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
+
+    @classmethod
+    def ndim_supp(cls, rng, weights, *components):
+        # We already checked that all components have the same dimensionality
```
I think that the signature of this method should be `ndim_supp(cls, weights, *components)` since an `rng` doesn't seem to be provided [here](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/distribution.py#L587)

ricardoV94 replied:
Good catch, I think you're right!

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,178 @@ class Mixture(Distribution):
+        # We return rv, which is a pseudo RandomVariable, that contains a choices RV
+        # in its inner graph. We want super().dist() to generate components + 1 rngs for
+        # us, and it will do so based on how many elements we return here
+        return (*rv.tag.components, rv)
```
Perhaps better to return Aesara `NoneConst` instead of `rv` to avoid confusion down the road

ricardoV94 commented:
Figured out the weights shape padding/broadcasting thanks tho @lucianopaz!

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,211 @@ class Mixture(Distribution):
-        super().__init__(shape, dtype, defaults=defaults, *args, **kwargs)
+        # Append  missing dimensions (if any) to the left of stacked_components
+        stacked_components_ = at.stack(components_, axis=-1)
+        stacked_components_padded_ = at.shape_padleft(
```
I think this is just the `weights_ndim_reps`

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,211 @@ class Mixture(Distribution):
+        # Extract support and replication ndims from components and weights
+        component_ndim_supp = components[0].owner.op.ndim_supp
+        component_ndim_reps = components[0].ndim - component_ndim_supp
+        weights_ndim_reps = max(0, weights.ndim - component_ndim_reps - 1)
```
We could instead change the size of the `Mixture` so that there are never batched dimensions coming exclusively from the weights. Perhaps issue a UserWarning before doing this...

ricardoV94 commented on pymc/distributions/mixture.py:
```diff
@@ -112,454 +132,211 @@ class Mixture(Distribution):
+            # Create new rng for the choices internal RV
+            mix_indexes_rng = aesara.shared(np.random.default_rng())
+
+        # Extract support and replication ndims from components and weights
```
Rename `replication/reps` to `batched`

ricardoV94 commented on pymc/tests/test_mixture.py:
```diff
@@ -627,53 +632,27 @@ def test_normal_mixture_nd(self, nd, ncomp):
-            # observed data.
-            # Furthermore, the Mixture will also raise errors when the observed
-            # data is multidimensional but it does not broadcast well with
-            # comp_dists.
```
This no longer seems to be a problem. @lucianopaz can you confirm that if the current tests pass, this is indeed fine?

ricardoV94 committed:
Refactor TestMixtureVsLatent

The two tests relied on implicit behavior of V3, where the dimensionality of the weights implied the support dimension of mixture distribution. This, however, led to inconsistent behavior between the random method and the logp, as the latter did not enforce this assumption, and did not distinguish if values were mixed across the implied support dimension.

In this refactoring, the support dimensionality of the component variables determines the dimensionality of the mixture distribution, regardless of the weights. This leads to consistent behavior between the random and logp methods as asserted by the new checks.

Future work will explore allowing the user to specify an artificial support dimensionality that is higher than the one implied by the component distributions, but this is for now not possible.

ricardoV94 committed:
Set `expand=True` when calling `change_size` in `SymbolicDistribution`

ricardoV94 committed:
Remove MixtureSameFamily

Behavior is now implemented in Mixture

ricardoV94 commented:
Except for the moment, everything seems ready for review. Thanks to both @lucianopaz and @larryshamalama for your inputs an help!

ricardoV94 committed:
Update release notes

larryshamalama commented on pymc/distributions/mixture.py:
```diff
@@ -112,457 +125,249 @@ class Mixture(Distribution):
+        if single_component:
+            # One dimension is taken by the mixture axis in the single component case
+            ndim_batch -= 1
+        weights_ndim_batch = max(0, weights.ndim - ndim_batch - 1)
```
A bit confused with everything going on with dimensions 😅 What is the range of `weights_ndim_batch` here? Is it either 0 or 1 (based on the `if` statement)?

ricardoV94 replied:
`weights_ndim_batch` is the extra dimensions besides what is implied by the components dimensions. It can be arbitrarily large.
```python
# In this example, weights_ndim_batch = 4
pm.Mixture(w=(np.full((4, 4, 4, 4, 4, 2), comp_dists=pm.Normal.dist(size=(4, 2)))`
```

ricardoV94 replied:
I added a comment to explain better in the code

larryshamalama commented on pymc/distributions/mixture.py:
```diff
@@ -112,457 +125,249 @@ class Mixture(Distribution):
+        # Index components and squeeze mixture dimension
+        if single_component:
+            mix_out_ = at.take_along_axis(components_[0], mix_indexes_padded, axis=-ndim_supp - 1)
+            # There is a Aeasara bug in squeeze with negative axis
```
After some thought, this sounds right!

larryshamalama commented on pymc/distributions/mixture.py:
```diff
@@ -112,457 +125,249 @@ class Mixture(Distribution):
+    if len(components) == 1:
+        # Need to broadcast value across mixture axis
+        mix_axis = -components[0].owner.op.ndim_supp - 1
+        components_logp = logp(components[0], at.expand_dims(value, mix_axis))
```
Looks good!

larryshamalama reviewed (commented):
Just left a single question, looks great :)

ricardoV94 committed:
Update Mixture docstrings

* Emphasize equivalency between iterable of components and single batched component
* Add example with mixture of two distinct distributions
* Add example with multivariate components

ricardoV94 committed:
Add warning when using iterable with single Mixture component

ricardoV94 commented:
@OriolAbril Any idea why the `tensor_like` here does not have a link to the glossary? https://pymc--5438.org.readthedocs.build/en/5438/api/distributions/generated/pymc.NormalMixture.html

OriolAbril commented:
There should be spaces both before and after the colon that separates parameter name and type

ricardoV94 committed:
Refactor NormalMixture

ricardoV94 commented:
Moments are also ready thanks to @larryshamalama. This is ready for review and merge :rocket:

larryshamalama commented on pymc/distributions/mixture.py:
```diff
@@ -53,516 +72,357 @@ class Mixture(Distribution):
+        components_ndim_supp = set()
+        for dist in comp_dists:
+            # TODO: Allow these to not be a RandomVariable as long as we can call `ndim_supp` on them
+            #  and resize them
```
Just wanting to clarify (again?) for my understanding: what else can `dist` be? For instance, is this where we want to be more flexible in allowing things like nested mixtures and censored distributions? E.g. something like [line 208 below](https://github.com/pymc-devs/pymc/pull/5438/files#diff-a46f0cbffb99eea1c2c7c51bfae351225ae8f38ad9b825d0b4687a7456bc6288R208)

ricardoV94 replied:
Yeah exactly!

larryshamalama reviewed (commented):
Looks great 🙂 Thanks for leading the effort @ricardoV94
